### ISSUE TITLE
lazy read disable for HTTP1 codec to reduce unnecessary syscalls

### ISSUE THREAD

**Issue by wbpcode** [envoyproxy/envoy#19900]
*Title*: *lazy read disable for HTTP1 codec to reduce unnecessary syscalls*

*Description*:

I get some perf result of latest Envoy by simple http1.1 bench. The `readDisable` calls from HTTP1 codec result in ~7% cpu overhead.

After checking related codes, I found that we always call the `readDiable(true)` to disable read event of downstream connection after receiving entire downstream request. And then call the `readDiable(false)` to re-enable read event of donwstream connection after complete downstream request.

```
// Disable read event of downstream connection after receiving entire downstream request.
ParserStatus ServerConnectionImpl::onMessageCompleteBase() {
  ASSERT(!handling_upgrade_);
  if (active_request_) {

    // The request_decoder should be non-null after we've called the newStream on callbacks.
    ASSERT(active_request_->request_decoder_);
    active_request_->response_encoder_.readDisable(true);
    active_request_->remote_complete_ = true;
    // ......
}
```

```
  // Re-enable read event of donwstream connection after complete downstream request.
  ~StreamEncoderImpl() override {
    // When the stream goes away, undo any read blocks to resume reading.
    while (read_disable_calls_ != 0) {
      StreamEncoderImpl::readDisable(false);
    }
  }
```

In this way, we can ensure that we only process one request at a time for the same downstream connection. However, most HTTP1 clients will not continue to send the next request until they have completed one. So in most cases, repeated `readDisable` calls are meaningless.

I think we could lazy call readDisable. `readDisable(true)` is only called when an active request has not yet completed and new data is received.

I wonder if this solution is possible? @KBaichoo @mattklein123 

![image](https://user-images.githubusercontent.com/12389633/153335488-d34a7140-4c7d-4d34-8457-800b925b0841.png)

**Comment by YaoZengzeng** [comment]
I've discussed with @alyssawilk about the `readDisable()` stuff, maybe it would be helpful. (ref: https://github.com/envoyproxy/envoy/pull/10978)

**Comment by wbpcode** [comment]
Thanks @YaoZengzeng.

**Comment by wbpcode** [comment]
My current simple solution is as this:

```
Http::Status ServerConnectionImpl::dispatch(Buffer::Instance& data) {
  if (active_request_ != nullptr && active_request_->remote_complete_) {
    active_request_->response_encoder_.readDisable(true);
    return okStatus();
  }
  return ConnectionImpl::dispatch(data);
}

```
`readDisable(true)` is only called when an active request has not yet completed and new data is received.

**Comment by alyssawilk** [comment]
I'd be inclined to go the other way.  simply don't read disable for HTTP when done with a request, but error out if getting more data when a response is not completed (as Envoy doesn't handle pipelined requests and basically no one sends them)

**Comment by wbpcode** [comment]
> I'd be inclined to go the other way. simply don't read disable for HTTP when done with a request, but error out if getting more data when a response is not completed (as Envoy doesn't handle pipelined requests and basically no one sends them)

Sounds ok for me.

**Comment by KBaichoo** [comment]
I think Envoy turns pipelined requests into serial requests e.g. we will do a read of the socket, and if it has multiple requests there, we'll parse the first one (read disable) and have the others buffered (but not consuming more than the network read size bytes). There a few tests in https://github.com/envoyproxy/envoy/blob/06505f2424ad3f894af404a07bcfd7582aa8f3b9/test/integration/integration_test.cc#L1047 that demonstrate this.

I think if we don't readDisable, we'll end up getting the file read event for the data in the network buffer. I'm not certain, but I don't think the codec will like that https://github.com/envoyproxy/envoy/blob/06505f2424ad3f894af404a07bcfd7582aa8f3b9/source/common/http/http1/codec_impl.cc#L1115

**Comment by wbpcode** [comment]
@KBaichoo Then would lazy read disable + manual read event (after one request complete and read buffer not empty) be feasible?

**Comment by KBaichoo** [comment]
Thanks for the code snippet. I'm leaning towards that approach as from the code snippet, 

I'd expect for pipelined requests, we'd end up getting worse perf as we'd get a socket event, dispatch that request, get the second socket event then readDisable. But as alyssa points out this is rarely used, so perhaps this reasonable.

For non-pipeline, we'd lazily readDisable clients, which, if the client didn't send additional data until it received the response would avoid the syscalls as you point out

**Comment by wbpcode** [comment]
closed by #20148
